Avoid negative integer overflow when filesize < io_->tell()

kevinbackhouse opened the pull request:
This fixes #791.

I added a new utility function to `BasicIO` to make it more convenient to read `rcount` bytes and throw an exception if not enough bytes are available. I called it `readOrThrow` because I couldn't think of a good name for it.

piponazo reviewed (commented):
Very nice idea! Ideally we could start replacing step by step the old usages of `read` by this new `readOrThrow` so that we are always sure about checking the whether the **read** operation succeed.

I would even vote for not having the 3rd argument and always throw the same kind of error (`Exiv2::kerCorruptedMetadata`) . 

I would also like to see @D4N 's opinion in this matter.

D4N commented:
I like this new API and will follow up with a proper review hopefully in the next few days. I have two comments:
- imho this is quite an API change for the 0.27 maintenance branch, not sure if we should introduce this
- `readOrThrow` should imho replace `read` completely, except for rare cases where we don't want an exception.

clanmills commented:
I agree with @D4N Please don't make API changes on 0.27-maintenance.

In fact, I should open an issue on myself to investigate if we made API changes in 0.27.1 and I'll update the release process to verify that we haven't introduced changes.  The aim of the v0.27-dots is to avoid API changes so security fixes can be provided by swapping DLLs without recompiling applications.

piponazo commented:
I also agree with @D4N comments:

1. I did not think about the API change, but it is true. We should not change the API in 0.27-maintenance.
2. We could keep the `read` function with the same signature but changing the implementation for always checking if the read was successful or not. In principle it should not break anything and it should make the code safer against some of these cooked POCs.

clanmills commented:
Can we use a macro or template to implement `readOrThrow()`, then we don't change the API?

kevinbackhouse committed:
Avoid negative integer overflow when `filesize < io_->tell()`.

This fixes #791.

kevinbackhouse commented:
I changed this to avoid the API change by making `readOrThrow` a static function in `webpimage.cpp`. I also posted #813, which adds the API change on master.

piponazo commented:
Hi @kevinbackhouse , as I commented in #813 the build on windows is not passing:

```
  C:\projects\exiv2-wutfp\src\webpimage.cpp(524): warning C4018: '<=': signed/unsigned mismatch [C:\projects\exiv2-wutfp\build\src\exiv2lib.vcxproj]
```
If you do not have a windows development environment let me know and I'll take a look at it.

kevinbackhouse committed:
Fix signed/unsigned mismatch.

kevinbackhouse commented:
@piponazo: I have pushed a commit to fix this. The problem is that `long` is only 32 bits on Windows. Which is a problem because it means that a `uint32_t` could overflow when cast to `long`. So I have changed all the types to `long` but also added extra checks for overflow.

D4N commented on src/webpimage.cpp:
```diff
@@ -489,15 +498,16 @@ namespace Exiv2 {
 
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
```
See my comment in #813, use `Safe::add`.

kevinbackhouse replied:
done

D4N commented on src/webpimage.cpp:
```diff
@@ -508,10 +518,12 @@ namespace Exiv2 {
+            readOrThrow(*io_, chunkId.pData_, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
+            readOrThrow(*io_, size_buff, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
+            const long size = Exiv2::getULong(size_buff, littleEndian);
+            enforce(0 <= size, Exiv2::kerCorruptedMetadata);
```
See my comment in #813, this is relying on undefined behavior. To properly check this, you must do something like this:
``` C++
const uint32_t size = Exiv2::getULong(size_buff, littleEndian);
enforce(size < std::numeric_limits<long>::max(), Exiv2::kerCorruptedMetadata);
const long long_size = size;
```

kevinbackhouse replied:
done

D4N reviewed (dismissed):
Overall this looks good to me for `0.27-maintenance`, I've added a few notes concerning the overflow checks.

D4N committed:
Update tests/bugfixes/github/test_issue_791.py

Co-Authored-By: kevinbackhouse <kev@semmle.com>

kevinbackhouse committed:
Improve portability of bounds checks.

kevinbackhouse committed:
Revert accidental whitespace change.

kevinbackhouse committed:
Small change to comment.
